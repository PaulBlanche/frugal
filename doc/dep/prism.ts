export { default } from "npm:prismjs@1.29.0";
export * from "npm:prismjs@1.29.0";

import "npm:prismjs@1.29.0/components/prism-jsx.js";
import "npm:prismjs@1.29.0/components/prism-typescript.js";
import "npm:prismjs@1.29.0/components/prism-tsx.js";
import "npm:prismjs@1.29.0/components/prism-json.js";
import "npm:prism-svelte@0.5.0";
