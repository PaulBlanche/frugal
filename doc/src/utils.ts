type ThrottleOptions = {
  timeout: number;
  leading?: boolean;
  trailing?: boolean;
};

export function throttle<ARGS extends any[]>(
  callback: (...args: ARGS) => void,
  { timeout, leading = false, trailing = false }: ThrottleOptions,
): (...args: ARGS) => void {
  let timeoutHandle: number | undefined = undefined;
  let previous = 0;

  return (...args) => {
    const now = Date.now();

    if (!previous && !leading) {
      previous = now;
    }

    const remaining = timeout - (now - previous);

    if (remaining <= 0) {
      clearTimeout(timeoutHandle);
      timeoutHandle = undefined;
      previous = now;
      callback(...args);
    } else if (timeoutHandle === undefined && trailing) {
      timeoutHandle = setTimeout(() => {
        previous = Date.now();
        timeoutHandle = undefined;
        callback(...args);
      }, remaining);
    }
  };
}
