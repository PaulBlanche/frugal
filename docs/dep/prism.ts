export { default } from 'https://esm.sh/prismjs@1.29.0';
export * from 'https://esm.sh/prismjs@1.29.0';

import 'https://esm.sh/prismjs@1.29.0/components/prism-jsx.js?no-check';
import 'https://esm.sh/prismjs@1.29.0/components/prism-typescript.js?no-check';
import 'https://esm.sh/prismjs@1.29.0/components/prism-tsx.js?no-check';
import 'https://esm.sh/prismjs@1.29.0/components/prism-json.js?no-check';
