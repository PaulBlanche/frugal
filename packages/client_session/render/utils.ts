import { NodeType } from './types.ts';

export function clone(node: Node) {
    if (isScriptElement(node)) {
        // for scripts cloning is not enough, we need to recreate the script
        // element for it to be loaded/evaluated again
        const script = document.createElement('script');
        for (const attribute of node.attributes) {
            script.setAttribute(attribute.name, attribute.value);
        }
        script.innerHTML = node.innerHTML;
        return script;
    }
    return node.cloneNode(true);
}

function isScriptElement(node: Node): node is HTMLScriptElement {
    return isElement(node) && node.nodeName === 'SCRIPT';
}

export function isElement(node: Node): node is Element {
    return node.nodeType === NodeType.ELEMENT_NODE;
}

export function hash(
    element: Element,
    attributeFilter?: string[],
): string {
    const key = [];

    if (attributeFilter) {
        for (const name of attributeFilter) {
            const value = element.getAttribute(name);
            if (value) {
                key.push(`${name}="${element.getAttribute(name)}"`);
            }
        }
    } else {
        for (const attribute of element.attributes) {
            key.push(`${attribute.name}="${attribute.value}"`);
        }
    }

    key.sort();

    return `${element.tagName} ${key.join(' ')} ${element.innerHTML}`;
}

type Attribute = [name: string, value: string | boolean];

const BOOLEAN_ATTRIBUTES = [
    'allowfullscreen',
    'async',
    'autofocus',
    'autoplay',
    'checked',
    'controls',
    'default',
    'defer',
    'disabled',
    'formnovalidate',
    'inert',
    'ismap',
    'itemscope',
    'loop',
    'multiple',
    'muted',
    'nomodule',
    'novalidate',
    'open',
    'playsinline',
    'readonly',
    'required',
    'reversed',
    'selected',
];

export function getAttributes(element: Element): Attribute[] {
    const attributes: Record<string, string | boolean> = {};

    for (const { name, value } of element.attributes) {
        attributes[name] = value;
    }

    for (const name of BOOLEAN_ATTRIBUTES) {
        attributes[name] = element.hasAttribute(name);
    }

    return Object.entries(attributes);
}
