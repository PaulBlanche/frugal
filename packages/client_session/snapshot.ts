import * as murmur from '../murmur/mod.ts';

type Node = {
    name: string;
    attributes: { name: string; value: string }[];
    content: string;
    hash: string;
    element: Element;
};

export type Head = Node[];

export function serialize(head: HTMLHeadElement) {
    const snapshot: Head = [];

    head.childNodes.forEach((node) => {
        if (!(node instanceof Element)) {
            return;
        }

        const hash = new murmur.Hash();

        const name = node.nodeName.toLowerCase();
        const content = node.textContent ?? '';

        const sorted: string[] = [];
        const attributes: { name: string; value: string }[] = [];
        for (let i = 0, length = node.attributes.length; i < length; i++) {
            const attribute = node.attributes[i];
            attributes.push({
                name: attribute.name,
                value: attribute.value,
            });
            sorted.push(`${attribute.name}="${attribute.value}"`);
        }

        hash.update(name);

        sorted.sort();
        sorted.forEach((value) => {
            hash.update(value);
        });

        hash.update(content);

        snapshot.push({
            name,
            attributes,
            content,
            hash: hash.digest(),
            element: node,
        });
    });

    return snapshot;
}
