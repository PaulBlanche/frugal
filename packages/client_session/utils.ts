export function isInternalUrl(url: URL) {
    const rootURL = new URL('/', document.baseURI);
    return url.href.startsWith(rootURL.href);
}

export function getClosestParentNavigableAnchor(
    target: EventTarget,
): HTMLAnchorElement | undefined {
    if (target instanceof Element) {
        const anchor = target.closest<HTMLAnchorElement>(
            'a[href]:not([target^=_]:not([download])',
        );
        if (anchor !== null) {
            return anchor;
        }
    }
}

export function getAnchorUrl(anchor: HTMLAnchorElement) {
    return new URL(anchor.href, document.baseURI);
}

const readyStateOrder: Record<DocumentReadyState, number> = {
    'loading': 0,
    'interactive': 1,
    'complete': 2,
};

declare global {
    interface WindowEventMap {
        'frugal:readystatechange': CustomEvent<
            { readystate: DocumentReadyState }
        >;
    }
}

export function onReadyStateChange(
    readyState: DocumentReadyState,
    callback: () => void,
) {
    if (readyStateOrder[document.readyState] >= readyStateOrder[readyState]) {
        callback();
    } else {
        document.addEventListener('readystatechange', () => {
            if (document.readyState === readyState) {
                callback();
            }
        });
    }

    addEventListener(
        'frugal:readystatechange',
        (event) => {
            if (event.detail.readystate === readyState) {
                callback();
            }
        },
    );
}
