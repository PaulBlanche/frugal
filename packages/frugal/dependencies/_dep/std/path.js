export * from "https://deno.land/std@0.203.0/path/mod.ts";
