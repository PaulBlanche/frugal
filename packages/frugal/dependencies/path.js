/** @type {import("./path/node.js")} */
let IMPLEMENTATION;

if (typeof Deno !== "undefined") {
    IMPLEMENTATION = await import("./path/deno.js");
} else {
    IMPLEMENTATION = await import("./path/node.js");
}

export const dirname = IMPLEMENTATION.dirname;
export const join = IMPLEMENTATION.join;
export const basename = IMPLEMENTATION.basename;
export const extname = IMPLEMENTATION.extname;
export const normalize = IMPLEMENTATION.normalize;
export const fromFileURL = IMPLEMENTATION.fromFileURL;
export const toFileURL = IMPLEMENTATION.toFileURL;
export const resolve = IMPLEMENTATION.resolve;
export const relative = IMPLEMENTATION.relative;
export const common = IMPLEMENTATION.common;
export const isAbsolute = IMPLEMENTATION.isAbsolute;
export const SEP = IMPLEMENTATION.SEP;
