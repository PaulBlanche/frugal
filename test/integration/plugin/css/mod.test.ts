import * as asserts from "../../../../dep/std/testing/asserts.ts";
import * as path from "../../../../dep/std/path.ts";
import * as snapshot from "../../../../dep/std/testing/snapshot.ts";
import { Config, context } from "../../../../mod.ts";
import { BuildHelper, setupFiles } from "../../../utils.ts";

if (import.meta.main) {
    const config = await loadConfig();
    await context(config).watch();
} else {
    await setupTestFiles();
}

Deno.test("css: build page with css dependencies", async (t) => {
    const config = await loadConfig();
    const helper = new BuildHelper(config);

    console.log(path.fromFileUrl(helper.config.rootdir));
    const watcher = Deno.watchFs(path.fromFileUrl(helper.config.rootdir), { recursive: true });

    const builderPromise = helper.build().then(() => {
        console.log("build done");
        watcher.close();
    });

    console.log("watch");

    for await (const event of watcher) {
        if (event.paths.some((path) => path.includes("/public/css/"))) {
            console.log(event);
            console.log(await Deno.readTextFile(event.paths[0]));
        }
    }

    snapshot.assertSnapshot(t, await Deno.readTextFile(new URL("css/page.css", helper.config.publicdir)));
});

Deno.test("css: css dependencies are watched", async () => {
    const config = await loadConfig();
    const helper = new BuildHelper(config);

    const context = helper.context();
    context.watch();

    await context.awaitNextBuild();

    const firstBuildCache = await context.cacheExplorer();
    const updatedAt = firstBuildCache.get("/page").updatedAt;

    // add a comment at the top of dep.css
    const depModuleURL = new URL("./project/dep.css", import.meta.url);
    const originalData = await Deno.readTextFile(depModuleURL);
    await Deno.writeTextFile(depModuleURL, `/*comment*/\n${originalData}`);

    await context.awaitNextBuild();

    const secondBuildChache = await context.cacheExplorer();
    asserts.assertNotEquals(updatedAt, secondBuildChache.get("/page").updatedAt);

    await Deno.writeTextFile(depModuleURL, originalData);

    await context.dispose();
});

async function setupTestFiles() {
    // clean everything from previous tests
    const base = new URL("./project/", import.meta.url);
    try {
        await Deno.remove(base, { recursive: true });
    } catch {}

    // setup clean files for current tests
    await setupFiles(base, {
        //####
        "./after.css": `.after {
            color: blue;
        }
`,

        //####
        "./after.ts": `import "./after.css";
`,

        //####
        "./before.css": `.before {
            color: blue;
        }
`,

        //####
        "./before.ts": `import "./before.css";
`,

        //####
        "./dep.css": `.dep {
            color: blue;
        }
`,

        //####
        "./main.css": `@import url(./dep.css);

.main {
    color: red;
}
`,

        //####
        "./page.ts": `import "./before.ts";
import "./main.css";
import "./after.ts";
import "https://meyerweb.com/eric/tools/css/reset/reset.css";
import "npm:prismjs@1.29.0/themes/prism.css";

export const pattern = "/page";

export function render() {
    return "";
}
`,

        //####
        "./frugal.config.ts": `import { Config } from "../../../../../mod.ts";
import { css } from "../../../../../plugins/css.ts";

export const config: Config = {
    self: import.meta.url,
    outdir: "./dist/",
    pages: ["./page.ts"],
    plugins: [css()],
    log: { level: "silent" },
};
`,
    });
}

async function loadConfig(): Promise<Config> {
    // load config busting deno cache
    const hash = String(Date.now());
    const { config } = await import(`./project/frugal.config.ts#${hash}`);
    return config;
}
