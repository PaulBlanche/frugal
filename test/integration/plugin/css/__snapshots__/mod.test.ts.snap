export const snapshot = {};

snapshot[`css: build page with css dependencies 1`] = `
"/* before.css */
" +
  ".before {
" +
  "  color: blue;
" +
  "}
" +
  "
" +
  "/* dep.css */
" +
  ".dep {
" +
  "  color: blue;
" +
  "}
" +
  "
" +
  "/* main.css */
" +
  ".main {
" +
  "  color: red;
" +
  "}
" +
  "
" +
  "/* after.css */
" +
  ".after {
" +
  "  color: blue;
" +
  "}
" +
  "
" +
  "/* https://meyerweb.com/eric/tools/css/reset/reset.css */
" +
  "html,
" +
  "body,
" +
  "div,
" +
  "span,
" +
  "applet,
" +
  "object,
" +
  "iframe,
" +
  "h1,
" +
  "h2,
" +
  "h3,
" +
  "h4,
" +
  "h5,
" +
  "h6,
" +
  "p,
" +
  "blockquote,
" +
  "pre,
" +
  "a,
" +
  "abbr,
" +
  "acronym,
" +
  "address,
" +
  "big,
" +
  "cite,
" +
  "code,
" +
  "del,
" +
  "dfn,
" +
  "em,
" +
  "img,
" +
  "ins,
" +
  "kbd,
" +
  "q,
" +
  "s,
" +
  "samp,
" +
  "small,
" +
  "strike,
" +
  "strong,
" +
  "sub,
" +
  "sup,
" +
  "tt,
" +
  "var,
" +
  "b,
" +
  "u,
" +
  "i,
" +
  "center,
" +
  "dl,
" +
  "dt,
" +
  "dd,
" +
  "ol,
" +
  "ul,
" +
  "li,
" +
  "fieldset,
" +
  "form,
" +
  "label,
" +
  "legend,
" +
  "table,
" +
  "caption,
" +
  "tbody,
" +
  "tfoot,
" +
  "thead,
" +
  "tr,
" +
  "th,
" +
  "td,
" +
  "article,
" +
  "aside,
" +
  "canvas,
" +
  "details,
" +
  "embed,
" +
  "figure,
" +
  "figcaption,
" +
  "footer,
" +
  "header,
" +
  "hgroup,
" +
  "menu,
" +
  "nav,
" +
  "output,
" +
  "ruby,
" +
  "section,
" +
  "summary,
" +
  "time,
" +
  "mark,
" +
  "audio,
" +
  "video {
" +
  "  margin: 0;
" +
  "  padding: 0;
" +
  "  border: 0;
" +
  "  font-size: 100%;
" +
  "  font: inherit;
" +
  "  vertical-align: baseline;
" +
  "}
" +
  "article,
" +
  "aside,
" +
  "details,
" +
  "figcaption,
" +
  "figure,
" +
  "footer,
" +
  "header,
" +
  "hgroup,
" +
  "menu,
" +
  "nav,
" +
  "section {
" +
  "  display: block;
" +
  "}
" +
  "body {
" +
  "  line-height: 1;
" +
  "}
" +
  "ol,
" +
  "ul {
" +
  "  list-style: none;
" +
  "}
" +
  "blockquote,
" +
  "q {
" +
  "  quotes: none;
" +
  "}
" +
  "blockquote:before,
" +
  "blockquote:after,
" +
  "q:before,
" +
  "q:after {
" +
  '  content: "";
' +
  "  content: none;
" +
  "}
" +
  "table {
" +
  "  border-collapse: collapse;
" +
  "  border-spacing: 0;
" +
  "}
" +
  "
" +
  "/* node_modules/.deno/prismjs@1.29.0/node_modules/prismjs/themes/prism.css */
" +
  "code[class*=language-],
" +
  "pre[class*=language-] {
" +
  "  color: black;
" +
  "  background: none;
" +
  "  text-shadow: 0 1px white;
" +
  "  font-family:
" +
  "    Consolas,
" +
  "    Monaco,
" +
  '    "Andale Mono",
' +
  '    "Ubuntu Mono",
' +
  "    monospace;
" +
  "  font-size: 1em;
" +
  "  text-align: left;
" +
  "  white-space: pre;
" +
  "  word-spacing: normal;
" +
  "  word-break: normal;
" +
  "  word-wrap: normal;
" +
  "  line-height: 1.5;
" +
  "  -moz-tab-size: 4;
" +
  "  -o-tab-size: 4;
" +
  "  tab-size: 4;
" +
  "  -webkit-hyphens: none;
" +
  "  -moz-hyphens: none;
" +
  "  -ms-hyphens: none;
" +
  "  hyphens: none;
" +
  "}
" +
  "pre[class*=language-]::-moz-selection,
" +
  "pre[class*=language-] ::-moz-selection,
" +
  "code[class*=language-]::-moz-selection,
" +
  "code[class*=language-] ::-moz-selection {
" +
  "  text-shadow: none;
" +
  "  background: #b3d4fc;
" +
  "}
" +
  "pre[class*=language-]::selection,
" +
  "pre[class*=language-] ::selection,
" +
  "code[class*=language-]::selection,
" +
  "code[class*=language-] ::selection {
" +
  "  text-shadow: none;
" +
  "  background: #b3d4fc;
" +
  "}
" +
  "@media print {
" +
  "  code[class*=language-],
" +
  "  pre[class*=language-] {
" +
  "    text-shadow: none;
" +
  "  }
" +
  "}
" +
  "pre[class*=language-] {
" +
  "  padding: 1em;
" +
  "  margin: .5em 0;
" +
  "  overflow: auto;
" +
  "}
" +
  ":not(pre) > code[class*=language-],
" +
  "pre[class*=language-] {
" +
  "  background: #f5f2f0;
" +
  "}
" +
  ":not(pre) > code[class*=language-] {
" +
  "  padding: .1em;
" +
  "  border-radius: .3em;
" +
  "  white-space: normal;
" +
  "}
" +
  ".token.comment,
" +
  ".token.prolog,
" +
  ".token.doctype,
" +
  ".token.cdata {
" +
  "  color: slategray;
" +
  "}
" +
  ".token.punctuation {
" +
  "  color: #999;
" +
  "}
" +
  ".token.namespace {
" +
  "  opacity: .7;
" +
  "}
" +
  ".token.property,
" +
  ".token.tag,
" +
  ".token.boolean,
" +
  ".token.number,
" +
  ".token.constant,
" +
  ".token.symbol,
" +
  ".token.deleted {
" +
  "  color: #905;
" +
  "}
" +
  ".token.selector,
" +
  ".token.attr-name,
" +
  ".token.string,
" +
  ".token.char,
" +
  ".token.builtin,
" +
  ".token.inserted {
" +
  "  color: #690;
" +
  "}
" +
  ".token.operator,
" +
  ".token.entity,
" +
  ".token.url,
" +
  ".language-css .token.string,
" +
  ".style .token.string {
" +
  "  color: #9a6e3a;
" +
  "  background: hsla(0, 0%, 100%, .5);
" +
  "}
" +
  ".token.atrule,
" +
  ".token.attr-value,
" +
  ".token.keyword {
" +
  "  color: #07a;
" +
  "}
" +
  ".token.function,
" +
  ".token.class-name {
" +
  "  color: #DD4A68;
" +
  "}
" +
  ".token.regex,
" +
  ".token.important,
" +
  ".token.variable {
" +
  "  color: #e90;
" +
  "}
" +
  ".token.important,
" +
  ".token.bold {
" +
  "  font-weight: bold;
" +
  "}
" +
  ".token.italic {
" +
  "  font-style: italic;
" +
  "}
" +
  ".token.entity {
" +
  "  cursor: help;
" +
  "}
"
`;
